cfdp-core/src/daemon.rs

```diff
@@ -511,127 +509,97 @@ impl<T: FileStore + Send + Sync + 'static> Daemon<T> {
-                                        panic!(
-                                            "Connection to Daemon Severed for Transaction {}",
-                                            transaction.id()
-                                        )
-                                    }
+                                }
+                                Command::Resume => transaction.resume()?,
+                                Command::Cancel => transaction.cancel()?,
+                                Command::Suspend => transaction.suspend()?,
+                                Command::Abandon => transaction.shutdown(),
+                                Command::Report(sender) => {
+                                    transaction.send_report(Some(sender))?
+                                }
+                                Command::Prompt(option) => {
+                                    transaction.prepare_prompt(option)
                                 }
                             }
+                        } else {
+                            panic!(
+                                "Connection to Daemon Severed for Transaction {}",
+                                transaction.id()
+                            )
                         }
-                    };
-                }
-                transaction.send_report(None)?;
-                Ok(id)
-            })?;
+                    },
+                    _ = tokio::time::sleep(timeout) => {
+                        transaction.handle_timeout()?;
+                    }
+                };
+            }
+            transaction.send_report(None)?;
+            Ok(id)
+        });
         Ok((id, transaction_tx, handle))
     }
 
     /// This function will consist of the main logic loop in any daemon process.
-    pub fn manage_transactions(&mut self) -> Result<(), Box<dyn std::error::Error + '_>> {
+    pub async fn manage_transactions(&mut self) -> Result<(), Box<dyn std::error::Error + '_>> {
         let mut sequence_num = self.sequence_num;
 
         // Create the selection object to check if any messages are available.
         // the returned index will be used to determine which action to take.
 
-        let mut selector = Select::new();
-        let transport = selector.recv(&self.transport_rx);
-        let user_primitive = selector.recv(&self.primitive_rx);
-
         // mapping of unique transaction ids to channels used to talk to each transaction
         let mut transaction_channels: HashMap<TransactionID, Sender<Command>> = HashMap::new();
 
         let mut cleanup = Instant::now();
 
         loop {
-            match selector.select() {
-                oper if oper.index() == transport => {
-                    match oper.recv(&self.transport_rx) {
-                        Ok(pdu) => {
-                            // find the entity this entity will be sending too.
+            select! {
```

Comment:
I know that the select! is a bit dangerous because  you can select from random futures which may do funny things when cancelled; but selecting from channels is one instance where select! makes sense because it looks very much like the original Unix select from sockets.

I haven't really used streams but I think you can merge them if they produce the same type. In our case the two channels produce different things (one produces PDU, the other one primitives) so I guess we would need first to make a wrapper (an enum). Looks more complicated to me...

Talking about readability, if it was up to me, I would move the body of those select branches into two functions handle_pdu and handle_primitive. Then the select would look like:

```
select! {
     pdu = self.transport_rx.recv() => {
             if let Some(pdu) = pdu {
                  handle_pdu(pdu).await;
             } else {
                  //transport channel closed
                   ....
              }
     },
      primitive = self.primitive_rx.recv() => {
             if let Some(primitive) = primitive {
                 hanlde_primitive(primitive).await;
             } else {
                  // user channel closed
                   ....
              }
       }
}
```

Reply:
Yeah you do end up needing to wrap in enums, then match the enum away. I can see why it can look more complicated sometimes. Definitely fine leaving it this way!

If you'd like to move some of the longer blocks into functions for readability I'm down for that. We can also make an issue and swing back to it. up to you!

Reply:
I think we can move later code around, I tried now to make minimum modifications for async.